### PR TITLE
Refactor all the RemotePossessionComponent tests

### PR DESCRIPTION
**Pull request by yujunting**
Refactor all the RemotePossessionComponent tests

Change all the APlayerController to AController.
If using APlayerController, it can not run all the tests together. And got a warning message and the test keeps alive until timeout.
`The following Workers failed to acknowledge FinishTest in time: [Client:1] `

Configured PR:https://github.com/spatialos/UnrealGDKTestGyms/pull/286

After changes, we remove the GameMode implementation, and we can put the Controller and Pawn to the position we want directly.
The successful test result:
image

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionController.cpp [hunk 1]
```diff
@@ -10,68 +10,78 @@
...
-void ATestPossessionPlayerController::OnUnPossess()
+void ATestPossessionController::OnUnPossess()
 {
 	Super::OnUnPossess();
-	UE_LOG(LogTestPossessionPlayerController, Log, TEXT("%s OnUnPossess()"), *GetName());
+	UE_LOG(LogTestPossessionController, Log, TEXT("%s OnUnPossess()"), *GetName());
 }
 
-void ATestPossessionPlayerController::RemotePossessOnClient_Implementation(APawn* InPawn, bool bLockBefore)
+void ATestPossessionController::ReleaseLock()
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionController.cpp:48]: Where is ReleaseLock called from?

Comment (yujunting) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionController.cpp:48]: Forgot to remove it.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp [hunk 2]
```diff
@@ -61,3 +76,28 @@ void ASpatialTestRemotePossession::AddWaitStep(const FWorkerDefinition& Worker)
 		WaitTime += DeltaTime;
 	});
 }
+
+void ASpatialTestRemotePossession::AddCleanStep()
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:80]: FYI there is a mechanism for reusing test steps called `FSpatialFunctionalTestStepDefinition`, see SpatialAuthorityTest for a usage example. You could define the Clean step as an `FSpatialFunctionalTestStepDefinition `and then use `AddStepFromDefinition` wherever the step is needed.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerMultiPossessionTest.cpp [hunk 3]
```diff
@@ -64,21 +62,30 @@ void ACrossServerMultiPossessionTest::PrepareTest()
...
-				if (FlowController->WorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Client)
+				ATestPossessionController* Controller = GetController();
+				if (Controller && Controller->HasAuthority())
 				{
-					ATestPossessionPlayerController* PlayerController = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
-					if (PlayerController && PlayerController->HasAuthority())
-					{
-						AssertTrue(PlayerController->IsMigration(), TEXT("PlayerController should migration"), PlayerController);
-					}
+					AssertTrue(Controller->IsMigration(), TEXT("Controller should migration"), Controller);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerMultiPossessionTest.cpp:75]:
```suggestion
					AssertTrue(Controller->IsMigration(), TEXT("Controller should have migrated"), Controller);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp [hunk 4]
```diff
@@ -44,17 +40,14 @@ void ACrossServerPossessionTest::PrepareTest()
...
+			if (Controller && Controller->HasAuthority())
 			{
-				ATestPossessionPlayerController* PlayerController = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
-				if (PlayerController && PlayerController->HasAuthority())
-				{
-					AssertTrue(PlayerController->HasAuthority(), TEXT("PlayerController should HasAuthority"), PlayerController);
-					AssertFalse(Pawn->HasAuthority(), TEXT("Pawn shouldn't HasAuthority"), Pawn);
-					PlayerController->RemotePossessOnServer(Pawn);
-				}
+				AssertTrue(Controller->HasAuthority(), TEXT("Controller should HasAuthority"), Controller);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:48]:
```suggestion
				AssertTrue(Controller->HasAuthority(), TEXT("Controller should have authority"), Controller);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp [hunk 5]
```diff
@@ -44,17 +40,14 @@ void ACrossServerPossessionTest::PrepareTest()
...
+				AssertFalse(Pawn->HasAuthority(), TEXT("Pawn shouldn't HasAuthority"), Pawn);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:49]:
```suggestion
				AssertFalse(Pawn->HasAuthority(), TEXT("Pawn shouldn't have authority"), Pawn);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp [hunk 6]
```diff
@@ -63,21 +56,29 @@ void ACrossServerPossessionTest::PrepareTest()
...
-				if (FlowController->WorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Client)
+				ATestPossessionController* Controller = GetController();
+				if (Controller && Controller->HasAuthority())
 				{
-					ATestPossessionPlayerController* PlayerController = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
-					if (PlayerController && PlayerController->HasAuthority())
-					{
-						AssertTrue(PlayerController->IsMigration(), TEXT("PlayerController should migration"), PlayerController);
-					}
+					AssertTrue(Controller->IsMigration(), TEXT("Controller should migration"), Controller);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:69]:
```suggestion
					AssertTrue(Controller->IsMigration(), TEXT("Controller should have migrated"), Controller);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp [hunk 7]
```diff
@@ -44,17 +40,14 @@ void ANoneCrossServerPossessionTest::PrepareTest()
...
+			if (Controller && Controller->HasAuthority())
 			{
-				ATestPossessionPlayerController* PlayerController = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
-				if (PlayerController && PlayerController->HasAuthority())
-				{
-					AssertTrue(PlayerController->HasAuthority(), TEXT("PlayerController should HasAuthority"), PlayerController);
-					AssertTrue(Pawn->HasAuthority(), TEXT("Pawn should HasAuthority"), Pawn);
-					PlayerController->RemotePossessOnServer(Pawn);
-				}
+				AssertTrue(Controller->HasAuthority(), TEXT("Controller should HasAuthority"), Controller);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp:48]:
```suggestion
				AssertTrue(Controller->HasAuthority(), TEXT("Controller should have authority"), Controller);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp [hunk 8]
```diff
@@ -44,17 +40,14 @@ void ANoneCrossServerPossessionTest::PrepareTest()
...
+				AssertTrue(Pawn->HasAuthority(), TEXT("Pawn should HasAuthority"), Pawn);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp:49]:
```suggestion
				AssertTrue(Pawn->HasAuthority(), TEXT("Pawn should have authority"), Pawn);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp [hunk 9]
```diff
@@ -63,21 +56,27 @@ void ANoneCrossServerPossessionTest::PrepareTest()
...
-				if (FlowController->WorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Client)
+				ATestPossessionController* Controller = GetController();
+				if (Controller && Controller->HasAuthority())
 				{
-					ATestPossessionPlayerController* PlayerController = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
-					if (PlayerController && PlayerController->HasAuthority())
-					{
-						AssertFalse(PlayerController->IsMigration(), TEXT("PlayerController shouldn't migration"), PlayerController);
-					}
+					AssertFalse(Controller->IsMigration(), TEXT("Controller shouldn't migration"), Controller);
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/NoneCrossServerPossessionTest.cpp:68]:
```suggestion
					AssertFalse(Controller->IsMigration(), TEXT("Controller shouldn't have migrated"), Controller);
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionController.h [hunk 10]
```diff
@@ -18,12 +18,14 @@ class ATestPossessionPlayerController : public APlayerController
...
+	UFUNCTION(CrossServer, Reliable)
+	void RemotePossess(APawn* InPawn);
 
-	UFUNCTION(Server, Reliable)
-	void RemotePossessOnClient(APawn* InPawn, bool bLockBefore);
+	void RemotePossessOnServer(APawn* InPawn, bool bLockBefore = false);
+
+	void ReleaseLock();
 
 	bool IsMigration() const { return BeforePossessionWorkerId != AfterPossessionWorkerId; }
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionController.h:30]: Maybe rename to HasMigrated()?

**Review by Vix2021 — APPROVED** [review]
Looks good, just added some minor feedback/suggestions.

**Comment by ImprobableNic** [comment]
I see that the main change is to stop using PlayerControllers in favor of Controllers because of some quirk of the testing framework.
This is unfortunate, because we are now testing a path that will be mostly irrelevant for users, who definitely want it to work with full PlayerController and Clients.
Maybe the coverage is enough, but this test now does not give a lot of confidence on it working in the real world.

Do you have more details about the test framework issue?
If it's just that there is something wrong with the test framework, I would rather move the tests to their own maps in order to keep a relevant scenario while we fix the test framework.

**Pull request closed without merging**